### DISCUSSION TITLE
WebSockets

### DISCUSSION THREAD

**Discussion by zubricks** [payloadcms/payload#4191] [Roadmap]
WebSockets are a powerful feature that we plan to bring to Payload. With WebSockets you’ll be able to subscribe to when documents are created, updated, or deleted over the same connection. 

This means you’d be able to build things like data driven dashboards, notification systems, and more.

EDIT: see [update below](https://github.com/payloadcms/payload/discussions/4191#discussioncomment-12280231).

**Comment by matthijs166** [comment]
Duplicate of #2534

Would love this feature!

**Comment by willviles** [comment]
Considering the move to Next.js and the likeliness many people will be deploying Payload in the same app as their frontend on Vercel (or other serverless environments), how would this work?

Would a WebSocket service be a standalone feature of Payload Cloud, perhaps?

**Reply by matthijs166** [reply]
https://youtu.be/zQdcAYmkvGs?t=731

**Comment by jnovak-SM2Dev** [comment]
Is there any timing on this? Just trying to plan for a project I have coming up in the next couple months that will need this feature.

**Comment by bormm** [comment]
@zubricks WebSockets are not a feature as described here, it's a protocol.
Therefor I wonder if the description is about supporting GraphQL subscriptions and of course support websockets as transport protocol for that. Otherwise, I could imagine that you want to reinvent the wheel and build some other notification or subscription mechanism in parallel to GraphQL that also uses WebSockets, but I don't think that's a good idea.

So please clarify, as its essential from my perspective that such a discussion uses the correct terms especially if a feature is flagged with priority 1.

**Comment by cgilly2fast** [comment]
Is there a branch that has been start for this feature? Looking to contribute to this feature

**Comment by carlosbensant** [comment]
PayloadCMS and [FeathersJS](http://feathersjs.com/) are my go-to Node.js frameworks, the last for Real-time Apps. I'd love to see how this will be implemented and the API you'll provide to interact with events and subscriptions.

**Comment by ccssmnn** [comment]
I'm having a good time using socket.io with payload. You can use Payloads local API in the socket.io handlers and the socket.io instance in Payload CMS hooks. With socket.io client libraries in your frontend, you can do whatever you want. With adapters, you can connect socket.io to your DB to scale this approach to multiple server instances, leveraging MongoDB change streams.

I wonder if that gets more complicated when moving to Payload CMS v3.

In my case, my client app needs to refresh its data on changes. So I emit poke events after relevant change or delete events. This was very straightforward, since socket.io is served from the same server as I can access the auth cookie. Here is the server-side setup.

```typescript
// example for hooks, that send poke events into organization rooms
import { extractId } from "./utils";
import { io } from "../../socket-io";
import {
  CollectionAfterChangeHook,
  CollectionAfterDeleteHook,
} from "payload/types";

export const sendOrganizationPokeAfterChange: CollectionAfterChangeHook =
  async ({ doc }) => {
    if (!doc.organization) return;
    const orgID = extractId(doc.organization);
    io.to(orgID).emit("poke");
  };

export const sendOrganizationPokeAfterDelete: CollectionAfterDeleteHook =
  async ({ doc }) => {
    if (!doc.organization) return;
    const orgID = extractId(doc.organization);
    io.to(orgID).emit("poke");
  };
```

```typescript
// setting up socket.io
import { Server as HTTPServer } from "node:http";
import payload from "payload";
import { Socket, Server as SocketIOServer } from "socket.io";
import { z } from "zod";
import { extractId } from "./utils";
import { createAdapter } from "@socket.io/mongo-adapter";
import { MongoClient } from "mongodb";

export let io: SocketIOServer;

export async function initSocketIO(server: HTTPServer) {
  io = new SocketIOServer(server, {
    /* options */
  });

  io.on("connection", hanleOnConnection);

  await initMongoDBConnection(io);

  return io;
}

/**
 * when the request comes from a logged in user,
 * add this socket into every organization room that
 * this user is member of
 */
async function hanleOnConnection(socket: Socket) {
  const { user } = await checkAuthCookie(socket.request.headers.cookie);
  if (!user) {
    socket.disconnect(true);
    return;
  }
  const organizationIDs = user?.organizations?.map((v) =>
    extractId(v.organization),
  );
  for (const orgID of organizationIDs ?? []) {
    socket.join(orgID);
  }
}

// this is where socket.io can leverage mongodb to support messaging accross multiple server instances
async function initMongoDBConnection(io: SocketIOServer) {
  const mongoClient = new MongoClient(process.env.DATABASE_URI as string);
  await mongoClient.connect();
  const mongoCollection = mongoClient
    .db()
    .collection("socket.io-adapter-events");
  await mongoCollection.createIndex(
    { createdAt: 1 },
    { expireAfterSeconds: 3600, background: true },
  );
  io.adapter(
    createAdapter(mongoCollection as any, {
      addCreatedAtField: true,
    }),
  );
}
```

```typescript
// when starting your server, attach socket.io to the http server
import { createServer } from "node:http";

import "dotenv/config";

import { initSocketIO } from "./socket-io";
import { initPayload } from "./payload";
import { initExpress } from "./express";

const main = async () => {
  const app = initExpress();
  const httpServer = createServer(app);
  await initPayload(app);
  initSocketIO(httpServer);
  httpServer.listen(3000);
};

main();
```

**Reply by carlosbensant** [reply]
Beautiful! Thanks for sharing!

**Reply by brankodencic** [reply]
This looks as a very nice solution. DId you try making a version of it for 3.0?

**Reply by ccssmnn** [reply]
I will migrate my project as soon as 3.0 is stable and share my solution

**Reply by cbratschi** [reply]
Next.js has no WebSocket support. There are a couple of tickets open and discussions. Next.js upgrades the HTTP connection itself and does not support a custom handling there.

I hope they will add WebSocket support to routes. Would be great to use export WS = message => ...

**Reply by pianomansam** [reply]
> Next.js has no WebSocket support. There are a couple of tickets open and discussions. Next.js upgrades the HTTP connection itself and does not support a custom handling there.
> 
> I hope they will add WebSocket support to routes. Would be great to use export WS = message => ...

The built-in server for Next.js does not, but a custom server allows you to. That is how trpc supports websockets w/ Next.

**Reply by cbratschi** [reply]
> > Next.js has no WebSocket support. There are a couple of tickets open and discussions. Next.js upgrades the HTTP connection itself and does not support a custom handling there.
> > I hope they will add WebSocket support to routes. Would be great to use export WS = message => ...
> 
> The built-in server for Next.js does not, but a custom server allows you to. That is how trpc supports websockets w/ Next.

Well, using a custom server is the current Payload 2.x setup to integrate it with Next.js but Payload 3.x is based on Next.js and a better solution for Websockets is needed without having to use custom servers.

Here is the Next.js discussion about WebSockets:

https://github.com/vercel/next.js/discussions/58698

Here is the example from TRPC for reference:

https://github.com/trpc/examples-next-prisma-websockets-starter/blob/main/src/server/prodServer.ts

**Comment by JeffreyArts** [comment]
@ccssmnn Have you already tried to upgrade your implementation to 3.0?

**Reply by ccssmnn** [reply]
i didn't see a satisfying way to create a socket.io instance, use nextjs with a custom server and make the socket.io instance available where i need it. you could deploy socket.io on the same nodejs server as your nextjs, but you need to treat it like an external server. nextjs code is bundled and there is no intended way of passing references to it.

currently migrating away from payload for my project 💔

**Reply by JeffreyArts** [reply]
I came across this article, which seems to be a solution, but I haven't tested it myself yet. https://medium.com/@kumarsehdev600/real-time-web-applications-with-next-js-and-socket-io-629706516597

I hope to find time for it in the next two weeks to give it a shot myself, if found. I'll let you know how it went.

**Reply by carlosbensant** [reply]
@ccssmnn What about using an external service like Pusher or Abby?

**Reply by ccssmnn** [reply]
@carlosbensant of course you can do that. My use case was relying on PayloadCMS as an application backend. I liked the idea of having a nodejs server where I can plug payload, socketio, as well as other frameworks like remix to have a reasonable setup.
It should be easily deployable on any setup, so clients can host it in their networks if they need to. Socket.io does even scale to a nodejs cluster. Websockets are not an option when deploying to vercel though.

Now that payload is so integrated with nextjs makes it easier to deploy wherever nextjs is deployable but cuts of my usecase.

**Reply by HarleySalas** [reply]
I exclusively host Next/Payload in standalone. In any case, the sort of thing I'd hope to see is some sort of adapter pattern, where realtime is optional and such adapter pattern should enable us to handle realtime in a similar way that storage providers do. We should be able to spin up a simple Hono instance, hook up party kit, or even things like Soketi, which can also be self-hosted.

Are there any ideas for a perfect scenario that fits inside of NextJS itself? I don't think so. I've already used Payload 3, with the Local API inside of a Hono application that uses the local api. It felt a little hacky, but perhaps a bit more time spent on it could yield a better result. It worked though. I've yet to experiment with things like Soketi, or Partykit, but I think it seems reasonable.

Most importantly, I just hope that realtime doesn't lock anyone in to having to host Nextjs a certain way, or that it doesn't force us to use any one particular provider.

**Reply by ccssmnn** [reply]
The problem with the local api inside another server is that you loose access to events of the nextjs payload that is connected to the (admin) UI.

**Reply by HarleySalas** [reply]
> The problem with the local api inside another server is that you loose access to events of the nextjs payload that is connected to the (admin) UI.

Ah, yeah, so anything cached would be out of sync basically, even though hooks generally should work otherwise?
Kinda forgot about that...

**Comment by GermanJablo** [comment]
🚨 UPDATE! 

I have started a PoC to query the DB in a reactive and real-time manner.
https://github.com/payloadcms/payload/pull/11334

It is still very immature, and it will probably take some time to arrive (no ETA yet).

I wanted to ask all of you who voted for this feature if something like this would solve what you're looking for.

I ask because this thread is called websocket, which seems to me to be a technical decision (in my PR I'm using Server Sent Events as an alternative). So, is what you want some real-time reactive way to access the DB, or is there some reason you want to instantiate and interact with a websocket server in a generic way?

**Reply by JeffreyArts** [reply]
I'd like to use websockets to develop flows like a chat functionality for example. I would have a chatMessages collection that would store all messages, and whenever something changes in the data (CRUD), it could inform the users via websockets. In the [plug-in](https://github.com/JeffreyArts/payload-plugin-socket-io) that I have been developing, the developer would use a custom socketAccess object that would fire SocketIO event emissions under the hood, to send messages to the connected clients (public, custom room & self). These events would be triggered at the same point in the application as the default access object is being handled

**Reply by GermanJablo** [reply]
For cases like these you could use the API I mentioned in the PR, and you wouldn't even have to know about the existence of WebSockets or SSE, which would be a matter of internal implementation.

**Reply by francesco-hayes** [reply]
Sadly, the problem with SSE are their limitations. SSE have a limit to active connections, number of events, and don't support sending binaries. For most people with smaller projects SEE will work just fine, but for production apps, with more than a couple hundred active users, they'll start hitting those limitations quickly on top of being limited to type of data being sent.

**Comment by ysfaran** [comment]
While investigating about real-time updates in Payload I stumbled across Convex BaaS.

Convex's real-time configuration and react hooks (`useQuery`/`useMutation`) look super clean to me. Some inspiration from Convex would bring Payload to another level in this regards.

**Comment by armelhbobdad** [comment]
Hi @GermanJablo. Please take a look #12506

**Comment by francesco-hayes** [comment]
I will say that I found a clean and non invasive way by using a separate service like Cloudflare Workers with Durable Objects to manage the WebSocket and WebHook side of things. This allows you to then send updates from within the Payload Collection Hooks to said service, which then sends the updates to the those connected to the WebSocket.

You'll have to connect to the WebSocket where needed instead of pulling the data from Payload directly. It's almost like putting in a middleman without loosing the benefit of Payload's collections.

**Reply by carlosbensant** [reply]
I have a similar implementation with **Redis pub/sub**. 

- PayloadCMS publishes changes through the `afterOperation` hook.
- WebSocket is subscribed to those changes.

But I would like to see your Cloudflare Workers impl.

**Comment by carlosbensant** [comment]
Trello Clone using PayloadCMS and Zero Sync
https://github.com/carlosbensant/trello-clone-payloadcms-zero-sync

Since PayloadCMS and Zero are betting on the Drizzle ORM, I didn't need much to make it work.

- [PayloadCMS generates the Drizzle schema built-in.](https://payloadcms.com/docs/database/postgres#options) You can even customize the file output generateSchemaOutputFile.
- You can easily use [drizzle-zero](https://github.com/BriefHQ/drizzle-zero) to generate the Zero Schema from the (Payload generated) Drizzle Schema.
- You just have to import the schema, add permissions, and the rest is history.

**Reply by matthijs166** [reply]
Oehh I love this!

**Reply by seanippolito** [reply]
hey this is cool, thanks

**Reply by seanippolito** [reply]
Digging into this more, if I add a pipeline through the Admin dashboard it is not displayed in the UI. If I add a pipeline through the frontend UI it works fine. Am I missing something?

I realize now that the data needs to be read and persisted to the Database and that appears to not be implemented in this application.

**Comment by dolanoadriano** [comment]
https://tanstack.com/blog/tanstack-db-0.1-the-embedded-client-database-for-tanstack-query

maybe it could be integrated with the new Tanstack DB?

**Comment by carlosbensant** [comment]
# Payload Sync Plugin

A real-time sync plugin for PayloadCMS that provides reactive queries (like Convex) and optimistic mutations.

## Features

- 🔄 **Real-time sync** - Automatic updates across all connected clients
- ⚡ **Optimistic updates** - Instant UI updates with server reconciliation
- 🎯 **Type-safe** - Full TypeScript support with PayloadCMS types
- 🚀 **Zero config** - Works out of the box with PayloadCMS
- 🔧 **Flexible** - Supports any backend architecture
- 🧠 [**Smart** Cross-Collection Updates](https://github.com/carlosbensant/payload-sync/blob/main/src/plugins/payload-sync/SMART_CROSS_COLLECTION_IMPLEMENTATION.md)
- 📌 **Persist active query subscriptions** after server restart
- ✨ **and many more** features

Link: https://github.com/carlosbensant/payload-sync/

**Reply by asafratzon** [reply]
Wow @carlosbensant this looks like some massive work. So first of all thanks for the sharing the plugin 🙌 

Couple of questions - I couldn't find any mention of "web sockets" in the plugin docs (md files) and wanted to be sure - is it using web sockets under the hood?

Another question is - given the magnitude of the plugin, is this something that's backed-up by payload's team as _the_ solution for realtime ? or is it an independent initiative?

---

Side note: personally, I don't understand why this discussion is called "Web sockets", I think it should have been named "Real-time API" or "Reactivity", mainly since "Web-sockets" is just one way to achieve reactivity, which in most cases is overly complex and has scaling limitations / complications. If the normal requirement of reactivity having the server notify the clients when data changes, then SSE would be a simpler solution.

**Reply by carlosbensant** [reply]
The plugin uses Server-Sent Events, but it was using Sockudo (an open-source alternative to Pusher) before adding the persistent active query subscription feature.

I wanted to see something like Convex (Reactive queries) in Payload, that's why I created this plugin. It is still a work in progress, but with testing, we can achieve a decent implementation.

**Comment by MurzNN** [comment]
I found a pretty easy way to implement WebSockets with Payload CMS, which supports Payload authorization!
1. Install the package https://npmjs.com/next-ws
```
pnpm add next-ws ws
```
2. Apply the patch for Next.js source (or better configure the package.json to automate this):
```
npx next-ws patch
```
3. Create a route to handle websocket connections - `src/app/api/ws/route.ts`
```ts
import config from '@payload-config'
import type { IncomingMessage } from 'node:http'
import { AuthResult } from 'node_modules/payload/dist/auth/operations/auth'
import { getPayload } from 'payload'

export function GET() {
  const headers = new Headers()
  headers.set('Connection', 'Upgrade')
  headers.set('Upgrade', 'websocket')
  return new Response('Upgrade Required', { status: 426, headers })
}

export function SOCKET(
  client: import('ws').WebSocket,
  request: import('http').IncomingMessage,
  server: import('ws').WebSocketServer,
  context: { params: Record<string, string | string[]> },
) {
  payloadAuthByHttpRequest(request).then(({ user }) => {
    // Put your websocket connection handler here.
  })
}

async function payloadAuthByHttpRequest(request: IncomingMessage): Promise<AuthResult> {
  // We need to convert Node.js headers to the Next.js headers format.
  const headers = new Headers()
  Object.entries(request.headers).forEach(([key, value]) => {
    headers.set(key, value as string)
  })
  const payload = await getPayload({ config })
  const authResult = await payload.auth({
    canSetHeaders: false,
    headers,
  })
  return authResult
}
```
And that's it!

The `payloadAuthByHttpRequest()` trick with headers is required to pass headers to the Payload.auth() in the Next.js format, because `next-ws` doesn't support it yet, but I posted an issue to fix it https://github.com/apteryxxyz/next-ws/issues/104 - please upvote! ;)

You can test the working implementation here https://github.com/MurzNN/payload-demo-chat/blob/main/src/app/api/ws/route.ts

**Reply by benny-sit** [reply]
The HMR is not working for the socket with this solution. There is already an issue about it in next-ws package  [https://github.com/apteryxxyz/next-ws/issues/74](apteryxxyz/next-ws/#74)

**Comment by MurzNN** [comment]
I went further and implemented a full Promise-based request-response connection over websockets, using https://github.com/bminer/ws-wrapper - will share the implementation a bit later.

**Reply by benny-sit** [reply]
Any progress?
There is some solution with ws packages only [https://github.com/aaronksaunders/payload-websockets-1-2025](https://github.com/aaronksaunders/payload-websockets-1-2025), but it is not supporting hmr...

**Reply by MurzNN** [reply]
The progress is (slowly) progressing and for now I managed to create something working here: https://github.com/MurzNN/payload-demo-chat - please check it out!

**Comment by thernstig** [comment]
I think focus should be to also include the WebTransport API in this, which has baseline support since March 2026. It will surely replace WebSockets more and more as time goes.

https://developer.mozilla.org/en-US/docs/Web/API/WebTransport_API
